src/System.Private.CoreLib/src/System/Environment.CoreRT.Unix.cs

```diff
@@ -13,22 +14,71 @@ namespace System
     public static partial class Environment
     {
         internal static int CurrentNativeThreadId => ManagedThreadId.Current;
-        
+
+        private static Dictionary<string, string> _environment;
+
         private static string GetEnvironmentVariableCore(string variable)
         {
             Debug.Assert(variable != null);
-            return Marshal.PtrToStringAnsi(Interop.Sys.GetEnv(variable));
+
+            var environment = LazyInitializer.EnsureInitialized(ref _environment, () => GetSystemEnvironmentVariables());
```

Comment:
CompareExchange that Jan suggested is less ceremony.

We are trading an extra delegate that we keep around forever for a saving of ~2 lines of code...